R/particle_filter_state_nested.R

```diff
@@ -83,31 +84,55 @@ particle_filter_state_nested <- R6::R6Class(
       }
 
       if (!is.null(initial)) {
-        initial_data <- initial(model$info(), n_particles, pars)
-        if (is.list(initial_data)) {
-          steps <- particle_steps(steps, initial_data$step)
-          model$set_state(initial_data$state, initial_data$step)
+        initial_data <- Map(initial,
+          info = model$info(),
+          pars = pars,
+          MoreArgs = list(n_particles = n_particles)
+        )
+        if (!is.null(unlist(initial_data))) {
+          if (any(c(c("step", "state") %in% names(initial_data[[1]])))) {
+            init_step <- unlist(lapply(initial_data, "[[", "step"))
+            if (length(init_step) != n_particles * length(pars)) {
+              if (length(init_step) == length(pars)) {
+                init_step <- rep(init_step, each = n_particles)
+              } else {
+                stop(sprintf("Expected 'step' to be length %d or %d",
```

Comment:
This error is probably not user-actionable; if the initial function gives unexpected outputs here it'll be hard to debug. So probably it needs adjusting so that it says something like: 

```
"initial() produced unexpected step length %d for parameter set %d (expected %d or %d)"
```

but then it might be easiest to manage that like:

```
initial_step_len <- lengths(init_step) # lengths is one of my fav underknown functions
if (length(unique(initial_step_len)) != 1) {
  stop("initial() produced unequal step lengths ...") # also provide some information about what was found
}
initial_step_len <- initial_step_len[[1]] # this is safe now because we know they're all exchangable
if (initial_step_len == 1) {
  # build your matrix by repetition
} else if (initial_step_len == n_particles) {
  # build your matrix by assembling it
} else {
  stop(sprintf("initial() produced unexpected step length %d (expected 1 or %d)",
               initial_step_len, n_particles))
}
```